### PR TITLE
Add stats_use_log2: return SDPA Stats (LSE) in base 2

### PR DESCRIPTION
**Pull request by YangXu1990uiuc**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes. (clang-format 21.1.6 and black `--line-length 160` on every touched file; the new test file carries the Apache-2.0 SPDX header.)
- [x] I reviewed the Hard Rules in the `AGENTS.md` for each directory this PR touches (see [root AGENTS.md § Reviewing a PR](https://github.com/NVIDIA/cudnn-frontend/blob/develop/AGENTS.md#reviewing-a-pr-human-or-agent)) and my changes comply, or I explain the exception below. (Append-only public signatures: the new parameter is last on `graph.sdpa()` and `sdpa_internal`. Rule S2: `Capabilities.stats_log2` is new, `SUPPORT_MATRIX_TRACKER.md` updated in the same commit. Kernel edits are single inline guarded statements, no helpers factored out of `@cute.kernel` bodies. Rule 5: see the stream fix below.)
- [x] I added GitHub labels: one `cat-*`, one or more `area:*` / `op:*`, and one `orig-*`.

## Affected area

- C++ frontend API or graph construction
- Python API or bindings
- FE OSS kernels or CuTeDSL

## Summary

`SDPA_attributes::set_stats_use_log2(bool)` / `graph.sdpa(..., stats_use_log2=False)`: when set, the forward `Stats` (LSE) output is returned in base 2, `max + log2(sum_exp)`, instead of the default natural-log `max + ln(sum_exp)`. Base-2 stats are exactly the natural-log stats times `log2(e)`; `O`, `score_max`, `score_sum_exp` and `sdpa_backward` (which keeps taking natural-log stats) are unchanged.

Who serves it:
- **FROST SDPA forward engines (all 7 specs: SM80, SM100 f16/fp8/mxfp8, SM107 fp8, SM120 f16/fp8), natively.** Every kernel already keeps the online softmax in the log2 domain and converts in the epilogue; each of the 16 prefill kernels gets one `cutlass.const_expr(STATS_LOG2)`-guarded `lse *= log2(e)` immediately before the final store, after the existing sink / dead-row(-inf) / seq-q-trim selects. Under split-KV the per-split partials stay natural and only the combine kernel's final LSE converts. The flag rides `TemplateParams` (compile-cache key), so base-2 is its own specialization.
- **cuDNN backend, `UNIFIED` implementation, cuDNN 9.28.0+**: the unified node sets `CUDNN_ATTR_OPERATION_SDPA_FWD_STATS_LOG2` (backend MR cudnn/cudnn!4402, draft). Older backends get `GRAPH_NOT_SUPPORTED` naming the version.
- **`COMPOSITE` declines** at expand time: the softmax op has no such attribute, and an appended pointwise on Stats has no servable engine (unified: "No valid engine configs"; composite runtime-compiled engines: NVRTC `COMPILATION_FAILED`) — that is what an earlier FE-only prototype of this PR hit on H100 and B200, and why the backend attribute exists.

Also in this PR, found while testing (Rule 5): `lower_dsl_prefill._execute` now runs the adapter's torch-side glue (O staging copy-back, dummy / carved-scratch fills) under `torch.cuda.stream(ExternalStream(handle_stream))`. With a handle on a non-default stream (the root `conftest.py` fixture) the copy-back could read `O_scratch` before the kernel wrote it; the natural-log graph failed 10-12/12 on B200 with `maxabs≈512` garbage in `O`, 0/12 after the fix. It was invisible before because the frost suite builds graphs without a handle. `bwd/engines.py` has the same `current_stream=` pattern and is not touched here.

Files: `graph_properties.h` (attribute, setter, JSON field), `scaled_dot_product_flash_attention.h` (unified attribute pass-through, composite decline), `python/pygraph/` (kwarg, appended last), `sdpa/graph_analyzer.py` (`has_stats_log2`), `sdpa/fwd/engines.py` (`Capabilities.stats_log2`, plumbing, stream fix), `sdpa/fwd/api_dsl.py`, `config_sm{80,100,120}.py`, 16 prefill kernels + `split_combine_sm100.py`, `SUPPORT_MATRIX_TRACKER.md`, `docs/operations/Attention.md`, tests.

## Why

FlashInfer's LSE convention is base 2 (FA2/FA3/TRT-LLM fold `log2(e)` into the softmax scale). Their cuDNN prefill backend converted cuDNN's natural-log stats with a Python-side multiply; when a second conversion landed inside the wrapper (flashinfer-ai/flashinfer#4663) the two stacked and every cuDNN LSE came out 1.44x too large while `O` stayed bit-exact — output-only tests stayed green and only an LSE oracle caught it after a rebase. LSE consumers (cascade / split-KV merges, DCP reductions, speculative decoding) are never in a single-op test, so a base drift surfaces far away as an accuracy problem. Owning the convention at the graph level removes the caller-side glue that can drift and the extra elementwise pass over the `(B, H, S_q, 1)` stats; in FROST and in the backend kernels it is a constant flip in an epilogue that was already log2-native.

## Related issues

Related to flashinfer-ai/flashinfer#4663. Backend counterpart: cudnn/cudnn!4402 (gitlab, draft).

## API and compatibility impact

- New optional attribute, default `false`; no behavior change unless set. Python kwarg appended last.
- `SDPA_attributes` JSON serialization gains `stats_use_log2`, as every prior attribute addition did.
- No new cuDNN version requirement for the default path. Base-2 on the cuDNN backend requires cuDNN 9.28.0+ and `UNIFIED`; FROST serves it on any backend.
- `sdpa_backward` continues to expect natural-log stats (documented); FROST backward engines are untouched.
- The random SDPA harness gains a `with_stats_log2` axis (`test_sdpa_random_bwd_L0`, `test_sdpa_random_bwd_ragged_L0`, 1-in-4): the forward writes base-2 stats, the harness snapshots them, restores the natural-log form in place before the backward (what a training caller would do) and compares the snapshot against `stats_ref * log2(e)`. `tools/cudnn_repro` carries `stats_use_log2` into that field.

## Testing

Random harness (`test_mhas_v2.py`, cuDNN 9.26 runtime so the base-2 cases are served by FROST only; without FROST they skip, they never fail):
- SM100 (B200), FROST on: `-k "test_sdpa_random_bwd_L0 and (test41 or ... or test49)"` → **46 passed, 53 skipped, 0 failed** over 99 cases; 24 drew `with_stats_log2=True`, of which 9 ran to completion with `'stats_log2' within limits` and 15 were waived by pre-existing "not supported backward graph" declines of the random shape (same waive rate as the unflagged cases). `bwd_L0 + bwd_ragged_L0` slice (26 cases): 10 passed / 16 skipped FROST on, 9 / 17 FROST off.
- Same 13-case bwd slice + `test_sdpa_stats_log2.py`, FROST on / off, per GPU on the box: H100 sm90 8p/30s / 8p/30s (no FROST fwd engine; base-2 skips), A100 sm80 21p/17s / 6p/32s, L40S sm89 6p/32s / 6p/32s, DRIVE P2021 sm100 22p/16s / 9p/29s. No failures anywhere.

Box: cudnn-dev-parley (B200 SM100, A100 SM80, H100 SM90), cuDNN 9.26.0 runtime (pip wheel), pybind module built from this worktree, `PYTHONPATH=build`. Commands run from `test/python` with `pytest -p no:cacheprovider --no-header -q`.

- `test_sdpa_stats_log2.py` (new; base-2 vs natural vs torch ref, AUTO/COMPOSITE/UNIFIED, causal/unmasked, fp16/bf16):
  - SM100, FROST disabled → **12 skipped** (9.26 backend has no attribute; unified/composite decline with a `GRAPH_NOT_SUPPORTED` naming cuDNN 9.28.0). Will exercise the backend path once cudnn/cudnn!4402 ships.
  - SM100, `CUDNN_FRONTEND_ENABLE_FROST_ENGINES=1`, together with `sdpa/frost/test_sdpa_graph_analyzer.py` → **114 passed**.
  - SM80, FROST on → **12 passed**.
- FROST LSE-vs-reference tests with the new `stats_use_log2` parametrization, `-m "L0 or L1"`:
  - SM100: `test_sdpa_fwd_dsl_sm100.py::test_dsl_sm100_strided_stats`, `test_sdpa_fwd_split_kv_sm100.py::test_combine_lse_matches_reference` (split 1 and 4), `test_sdpa_fwd_fp8_sm100.py` + `test_sdpa_fwd_mxfp8_sm100.py -k "stats or lse or combine"` → **63 passed**.
  - SM80: `test_sdpa_sm80_frontend_integration.py -k "stats or end_to_end"` → **6 passed**.
- Regression smoke on the default (natural-log) path, H100: `test_mhas_v2.py -k "(fwd_L0 or bwd_L0) and (test500 or test501 or test502 or test503)"` → **7 passed, 5 skipped** (skips are pre-existing backend not-supported cases).
- Not executed here: SM120/SM121 and SM107 kernels (`prefill_f16_sm120.py`, `prefill_fp8_sm120.py`, `prefill_d128_fp8_sm107.py`, `test_dsl_sm120_stats`) received the same one-line epilogue edit but this box has no such GPU — please run the sm120 lane in CI.
- Stream fix evidence: with the root-conftest handle (non-default stream) the natural-log graph on B200 failed 10-12/12 (`maxabs≈512` in O) before the change and 0/12 after; `test_sdpa_stats_log2.py` is the first test that reaches the FROST fwd path through a handle-bearing graph.
- `pre-commit run` (clang-format 21.1.6, black 160, SPDX) passed on commit.

---
<sub>note to self: claude::ad57a85b-a5a3-42f6-b37f-e3babc4f94f2 — "cuDNN SDPA prefill LSE convention" · cwd /home/scratch.yanxu_libs/cudnn_frontend · workspace /home/scratch.yanxu_libs/cudnn_frontend_stats_log2 (worktree on gh/develop); backend worktree /home/scratch.yanxu_gpu/cudnn_stats_log2; probes and logs in the session scratchpad</sub>

## Summary by CodeRabbit

* **New Features**
  * Added an optional `stats_use_log2` setting for FP16/BF16 SDPA forward operations.
  * Statistics can now use base-2 log-sum-exp values while attention outputs remain unchanged.
  * The option is available through C++ and Python APIs and is supported across applicable FROST and UNIFIED implementations.

* **Bug Fixes**
  * Preserved natural-log statistics requirements for backward operations.
  * Added clear unsupported-status handling where the selected implementation or cuDNN version cannot provide base-2 statistics.

* **Documentation**
  * Documented defaults, conversions, compatibility requirements, and implementation support.

### CODE REVIEW

**Comment by YangXu1990uiuc** [comment]
Refreshed on develop `e45ac96b6d61e50b9dca48eadc97d060fb4c1944` and rebuilt the Python binding from this PR against compatible local cuDNN development headers and upstream-pinned DLPack 1.3.

Fresh validation with Torch 2.13.0+cu130 and cuDNN backend 9.26:

```bash
python -m pytest -q test/python/test_sdpa_stats_log2.py
python -m pytest -q -m 'L0 or L1' \
  test/python/sdpa/frost/test_sdpa_fwd_dsl_sm100.py \
  test/python/sdpa/frost/test_sdpa_fwd_split_kv_sm100.py \
  -k 'strided_stats or combine_lse_matches_reference'
```

- Public Stats-log2 tests: **12 passed on SM100 and 12 passed on A100**.
- SM100 strided-Stats and split-KV/LSE regressions: **9 passed**.
- Pinned pre-commit hooks passed for all changed files.

These tests compare Stats to both natural-log graphs and independent references, and also check O. FROST was opted in. This does not claim that the cuDNN 9.28+ unified backend path, SM107, or SM120 was freshly validated; the backend counterpart and those target lanes remain separate merge/CI checks. Draft status is preserved. No new performance result.

@cudnn-ci-bot run python_tests

**Review by vedaanta — APPROVED** [review]
